client/hotrod/src/test/java/org/infinispan/hotrod/HotRodAsyncCacheTest.java

```diff
@@ -1,34 +1,286 @@
+   public void testPutAllGetAndRemoveAll(KeyValueGenerator<K, V> kvGenerator) {
+      Map<K, V> entries = new HashMap<>();
+      SubmissionPublisher<CacheEntry<K, V>> entriesPublisher = new SubmissionPublisher<>();
+      final CacheWriteOptions options = CacheWriteOptions.writeOptions()
+            .timeout(Duration.ofSeconds(15))
+            .lifespanAndMaxIdle(Duration.ofSeconds(20), Duration.ofSeconds(25))
+            .build();
+      CompletionStage<Void> putAll = cache.putAll(entriesPublisher, options);
+
+      for (int i = 0; i < 10; i++) {
+         final K key = kvGenerator.generateKey(cacheName, i);
+         final V value = kvGenerator.generateValue(cacheName, i);
+         entries.put(key, value);
+         entriesPublisher.submit(new CacheEntryImpl<>(key, value, null));
+      }
+
+      entriesPublisher.close();
+      await(putAll);
+
+      Flow.Publisher<CacheEntry<K, V>> publisher = cache.getAndRemoveAll(entries.keySet());
+      Map<K, CacheEntry<K, V>> retrieved = FlowUtils.blockingCollect(publisher)
+            .stream().collect(Collectors.toMap(CacheEntry::key, e -> e));
+
+      assertEquals(entries.size(), retrieved.size());
+      final CacheEntryVersion cve = new CacheEntryVersionImpl(0);
+      MapKVHelper<K, V> helper = new MapKVHelper<>(entries, kvGenerator);
+      for (Map.Entry<K, CacheEntry<K, V>> entry : retrieved.entrySet()) {
+         V expected = helper.get(entry.getKey());
+         assertNotNull(expected);
+         assertEntry(entry.getKey(), expected, kvGenerator, entry.getValue(), options, cve);
+      }
+
+      for (Map.Entry<K, V> entry : entries.entrySet()) {
+         assertAwaitEquals(null, cache.get(entry.getKey()));
+      }
+
+      // Removing keys that dont exists returns nothing.
+       retrieved = FlowUtils.blockingCollect(cache.getAndRemoveAll(entries.keySet()))
+            .stream().collect(Collectors.toMap(CacheEntry::key, e -> e));
+      assertTrue(retrieved.isEmpty());
+   }
+
+   @MethodSource("parameterized")
+   @ParameterizedTest(name = "testReplace[{0}]")
+   public void testReplace(KeyValueGenerator<K, V> kvGenerator) {
+      final K key = kvGenerator.generateKey(cacheName, 0);
+      final V initialValue = kvGenerator.generateValue(cacheName, 0);
+      final CacheWriteOptions options = CacheWriteOptions.writeOptions()
+            .timeout(Duration.ofSeconds(15))
+            .lifespanAndMaxIdle(Duration.ofSeconds(20), Duration.ofSeconds(25))
+            .build();
+
+      // Returns false for an unexistent entry.
+      final CacheEntryVersion cve0 = new CacheEntryVersionImpl(0);
+      assertAwaitEquals(false, cache.replace(key, initialValue, cve0));
+      assertEntry(key, null, kvGenerator, await(cache.put(key, initialValue, options)));
+      assertEntry(key, initialValue, kvGenerator, await(cache.getEntry(key)));
+
+      final V replaceValue = kvGenerator.generateValue(cacheName, 1);
+      final CacheEntryVersion cve2 = new CacheEntryVersionImpl(2);
```

Comment:
I am missing something, shouldn't the version be 1?

Reply:
I was expecting that too. I did not check into it, but line 265, with the first replace that returns false, seems to increase the version. Should we register a JIRA or something?

Reply:
That seems wrong to increment the version if a conditional operation failed. I would say that is a bug.

Reply:
After looking into it, the behavior is expected:

https://github.com/infinispan/infinispan/blob/d855ad4a2ed126ccc897989e6492039ad5dc1f54/server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java#L288-L294

Reply:
Alright, I guess technically it doesn't matter as long as they are unique.